cpp-client/deephaven/dhcore/include/private/deephaven/dhcore/immerutil/immer_column_source.h

```diff
@@ -41,59 +40,81 @@ struct ImmerColumnSourceImpls {
+    // pair of [begin, end) Immer iterators. But then, rather than store into that iterator range
+    // directly, those Immer iterators are passed to immer::for_each_chunk. This breaks down the
+    // Immer range into subranges of plain data, and invokes the copy_data_inner lambda. Then,
+    // 'copy_data_inner' just copies data in the normal C++ way.
+    auto copy_data_inner = [&dest_datap](const T *src_beginp, const T *src_endp) {
+      for (const T *current = src_beginp; current != src_endp; ++current) {
+        *dest_datap++ = *current;
       }
     };
 
-    auto copy_outer = [&src_data, src_null_flags, dest_nullp, &copy_data_inner,
-        &copy_nulls_inner](uint64_t src_begin, uint64_t src_end) {
+    auto copy_data_outer = [&src_data, &copy_data_inner](uint64_t src_begin, uint64_t src_end) {
       auto src_beginp = src_data.begin() + src_begin;
       auto src_endp = src_data.begin() + src_end;
       immer::for_each_chunk(src_beginp, src_endp, copy_data_inner);
+    };
+
+    rows.ForEachInterval(copy_data_outer);
+
+    // If the caller has opted out of getting null flags, we are done.
+    if (optional_dest_null_flags == nullptr) {
+      return;
+    }
 
-      if constexpr(!deephaven::dhcore::DeephavenTraits<T>::kIsNumeric) {
-        if (dest_nullp != nullptr) {
-          auto nulls_begin = src_null_flags->begin() + src_begin;
-          auto nulls_end = src_null_flags->begin() + src_end;
-          immer::for_each_chunk(nulls_begin, nulls_end, copy_nulls_inner);
+    // Otherwise (if the caller wants null flags), we do a similar algorithm to copy null flags.
+    // The one complication is that the column source only stores null flags explicitly for
+    // non-numeric types. For numeric types, the column source uses the Deephaven convention
+    // for nullness. To handle this, we have two different forms of the operation,
+    // one which supports the numeric convention and one which supports the non-numeric convention.
+    auto *dest_nullp = optional_dest_null_flags->data();
+
+    if constexpr (kTypeIsNumeric) {
+      auto copy_nulls_inner = [&dest_nullp](const T *data_begin, const T *data_end) {
+        for (const T *current = data_begin; current != data_end; ++current) {
+          auto is_null = *current == deephaven::dhcore::DeephavenTraits<T>::kNullValue;
+          *dest_nullp++ = is_null;
         }
-      } else {
-        // avoid clang complaining about unused variables.
-        (void)src_null_flags;
-        (void)dest_nullp;
-        (void)copy_nulls_inner;
-      }
-    };
-    rows.ForEachInterval(copy_outer);
+      };
+
+      auto copy_nulls_outer = [&src_data, src_null_flags, &copy_nulls_inner](uint64_t src_begin,
+          uint64_t src_end) {
+        auto src_beginp = src_data.begin() + src_begin;
+        auto src_endp = src_data.begin() + src_end;
+        immer::for_each_chunk(src_beginp, src_endp, copy_nulls_inner);
+      };
+      rows.ForEachInterval(copy_nulls_outer);
+    } else {
```

Comment:
Since the function is about to return right after the else anyway, can you consider short-circuiting this else by returning in the body of the if above, and then removing the else?

Looking at the actual file is going to be easier than trying to follow the logic in the diff, but in either case, I find that is easier for me to read code when you allow me to drop stacks of context earlier.

Reply:
We discussed on slack, it is an `if constexpr` so my comment doesn't apply/work.